File: day02/mvZeroes/moveZeroes.go
/*
给定一个数组 nums，编写一个函数将所有 0 移动到数组的末尾，同时保持非零元素的相对顺序。

示例:

输入: [0,1,0,3,12]
输出: [1,3,12,0,0]
说明:

必须在原数组上操作，不能拷贝额外的数组。
尽量减少操作次数。

*/

package main

import "fmt"

func moveZeroes(nums []int) {
	j := 0
	for i := 0; i < len(nums); i++ {
		if nums[i] != 0 {
			nums[j] = nums[i]
			if i != j {
				nums[i] = 0
			}
			j++
		}
	}
}

// 方法2 的好处在于，只会在nums[i] != 0 的时候才会进行交换
// 隔了一段时间再看这个题认识更加深刻
func moveZeroes2(nums []int) {
	j := 0
	for i := 0; i < len(nums); i++ {
		if nums[i] != 0 {
			nums[j], nums[i] = nums[i], nums[j]
			j++
		}
	}
}

func moveZeroes3(nums []int) {
	j := 0
	for i:=0;i<len(nums);i++ {
		if nums[i] != 0 {
			nums[j], nums[i] = nums[i], nums[j]
			j++
		}
	}
}

func moveZeroes4(nums []int) {
	j := 0
	for i:=0;i<len(nums);i++ {
		if nums[i] == 0 {
			continue
		}
		nums[i], nums[j] = nums[j], nums[i]
		j++
	}
}

func moveZeroes5(nums []int) {
	j := 0
	for i:=0;i<len(nums);i++ {
		if nums[i] ==0 {
			continue
		}
		nums[i], nums[j] = nums[j], nums[i]
		j++
	}
}

func main() {
	nums := []int{0, 1, 0, 3, 12}
	moveZeroes5(nums)
	fmt.Println(nums)
}

File: day02/mvZeroes/moveZeroes.rs
use std::collections::VecDeque;

fn move_zeroes(nums: &mut Vec<i32>) {
    let mut buf = VecDeque::new();
    let mut i = 0;
    for n in nums.to_owned() {
        if n == 0 {
            buf.push_back(i);
        } else {
            if buf.len() > 0 {
               if let Some(t) = buf.pop_front() {
                   nums.swap(t, i);
                   buf.push_back(i);
               } 
            }
        }
        i += 1;
    }
}

fn move_zeroes2(nums: &mut Vec<i32>) {
    let mut a = VecDeque::new();
    let mut j = 0;
    for n in nums.to_owned() {
        if n == 0 {
            a.push_back(i);
        } else {
            if a.len() > 0 {
                if let Some(t) = a.pop_front() {
                    nums.swap(t, i);
                    a.push_back(i);
                }
            }
        }
        j+=1;
    }
}


// 这个的思路和go 中的那个代码的思路是类似的
// ```
//func moveZeroes5(nums []int) {
//   	j := 0
//   	for i:=0;i<len(nums);i++ {
//   		if nums[i] ==0 {
//   			continue
//   		}
//   		nums[i], nums[j] = nums[j], nums[i]
//   		j++
//   	}
//   }
// ```
fn move_zeroes3(nums: &mut Vec<i32>) {
    let mut j = 0;
    for i in 0..nums.len() {
        if nums[i] == 0 {
            continue
        }
        let tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;

        j+=1;
    }
}

fn main() {
    let mut aa = vec![0,1,0,3,12];
    move_zeroes3(&mut aa);
    println!("{:?}", aa);
}